File: #96_Unique_Binary_Search_Trees/Solution.java
/* time: O(n^2) */

public class Solution {
    public int numTrees(int n) {
    	if (n == 0) return 1;
        int[] dp = new int[n + 1];
        dp[0] = 1;
        for (int i = 1; i <= n; i++) {
        	for (int j = 1; j <= i; j++) {
        		dp[i] += dp[j - 1] * dp[i - j];
        	}
        }
        return dp[n];
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	System.out.println(sol.numTrees(3));
    }
}

File: #96_Unique_Binary_Search_Trees/Solution.py
'''
Time limit exceed
'''

class Solution:
    # @param {integer} n
    # @return {integer}
    def numTrees(self, n):
        if n < 1:
        	return 0
        return self.countTree([x for x in range(1, n + 1)])

    def countTree(self, nums):
    	if not nums:
    		return 1
    	ans = 0
    	for x in range(len(nums)):
    		ans += self.countTree(nums[:x]) * self.countTree(nums[x + 1:])
    	return ans

sol = Solution()
print sol.numTrees(3)